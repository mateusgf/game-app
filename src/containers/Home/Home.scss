.home {
  &--btns-wrp{
    margin-top: 20px;
  }

  &--or-label {
    display: block;
    margin-top: 20px;
    margin-bottom: 20px;
    text-align: center;
    font-size: 1.65rem;
    font-weight: 700;
  }
}