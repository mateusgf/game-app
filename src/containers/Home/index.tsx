import { useState } from "react";
import "./Home.scss";
import Button from "../../components/Atom/Button";
import NewGameForm from "../../components/Molecule/NewGameForm";
import JoinGameForm from "../../components/Molecule/JoinGameForm";
import { routeCodes } from "../../constants/routes";
import { useHistory } from "react-router-dom";
import { useGame } from "../../context/Game";
import { usePlayer } from "../../context/Player";
import getPlayerByNickname from "../../api/requests/player/getPlayerByNickname";

const Home = () => {
  const history = useHistory();
  const { createPlayer } = usePlayer();
  const { createGame, joinGame } = useGame();
  const [isNewGameFormVisible, setIsNewGameFormVisible] = useState(false);
  const [isJoinGameFormVisible, setIsJoinGameFormVisible] = useState(false);

  const goToGamesList = () => {
    history.push(routeCodes.GAMES_LIST);
  };

  const newGameHandler = async (values: { nickname: string; numberOfRounds: string }) => {
    const player = await createPlayer(values.nickname);
    if (player && player.nickname) {
      const game = await createGame(player.nickname, values.numberOfRounds);
      if (game.id) {
        history.push({
          pathname: routeCodes.GAME.replace(":id", game.id),
        });
      }
    }
  };

  const joinGameHandler = async (values: { nickname: string; gameId: string }) => {
    const submitedPlayer = await getPlayerByNickname(values.nickname);

    // @TODO: check if game exists before creating the player

    if (!submitedPlayer || !submitedPlayer.nickname) {
      // Player does not exist yet, so create one
      await createPlayer(values.nickname);
    }

    if (await joinGame(values.gameId, values.nickname)) {
      history.push({
        pathname: routeCodes.GAME.replace(":id", values.gameId),
      });
    } else {
      // @TODO: threat error better here
      alert("Could not join the game");
    }
  };

  return (
    <div className="home">
      <h1>Rock, Paper & Scissors</h1>
      {isNewGameFormVisible ? (
        <>
          <h3>New game</h3>
          <NewGameForm onSubmit={newGameHandler} />
        </>
      ) : null}
      {isJoinGameFormVisible ? (
        <>
          <h3>Join game</h3>
          <JoinGameForm onSubmit={joinGameHandler} />
        </>
      ) : null}

      <div className="home--btns-wrp">
        <Button
          className="game-button--full-width"
          data-testid="btn-start-game"
          onClick={() => {
            setIsNewGameFormVisible(!isNewGameFormVisible)
            setIsJoinGameFormVisible(false);
          }}
        >
          Start Game
        </Button>

        <div className="home--or-label">or</div>

        <Button
          className="game-button--full-width"
          data-testid="btn-join-game"
          onClick={() => {
            setIsJoinGameFormVisible(!isJoinGameFormVisible)
            setIsNewGameFormVisible(false)
          }}
        >
          Join a game
        </Button>

        <div className="home--or-label">or</div>

        <Button className="game-button--full-width" data-testid="goto-listing" onClick={goToGamesList}>See started games</Button>
      </div>

    </div>
  );
};

export default Home;
