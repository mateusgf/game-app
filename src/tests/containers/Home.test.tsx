import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom";
import Home from "../../containers/Home";

describe("Home container", () => {
  test("match snapshot", () => {
    const { container } = render(<Home />);
    expect(container).toMatchSnapshot();
  });

  test("renders home", () => {
    const { container } = render(<Home />);
    const homeTitle = screen.getByText(/Rock, Paper & Scissors/i);
    const btnStartGame = screen.getByTestId("btn-start-game");
    const btnJoinGame = screen.getByTestId("btn-join-game");
    expect(btnJoinGame).toBeInTheDocument();
    expect(btnStartGame).toBeInTheDocument();
    expect(homeTitle).toBeInTheDocument();
  });

  test("renders new game form", async () => {
    const { container } = render(<Home />);
    fireEvent(
      screen.getByTestId("btn-start-game"),
      new MouseEvent("click", {
        bubbles: true,
        cancelable: true,
      }),
    );

    const form = await screen.findAllByTestId("new-game-form");
    const inputNickname = await screen.findAllByTestId("input-nickname");
    const inputNumberOfRunds = await screen.findAllByTestId("input-number-of-rounds");
    const btnStart = await screen.findAllByTestId("btn-start");
    expect(form).toBeTruthy();
    expect(inputNickname).toBeTruthy();
    expect(inputNumberOfRunds).toBeTruthy();
    expect(btnStart).toBeTruthy();
  });

  test("renders join game form", async () => {
    const { container } = render(<Home />);

    fireEvent(
      screen.getByTestId("btn-join-game"),
      new MouseEvent("click", {
        bubbles: true,
        cancelable: true,
      }),
    );

    const form = await screen.findAllByTestId("join-game-form");
    const inputNickname = await screen.findAllByTestId("input-nickname");
    const inputNumberOfRunds = await screen.findAllByTestId("input-game-id");
    const btnStart = await screen.findAllByTestId("btn-join");
    expect(form).toBeTruthy();
    expect(inputNickname).toBeTruthy();
    expect(inputNumberOfRunds).toBeTruthy();
    expect(btnStart).toBeTruthy();
  });
});
