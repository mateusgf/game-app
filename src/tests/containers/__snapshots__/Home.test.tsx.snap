// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home container match snapshot 1`] = `
<div>
  <div
    class="home"
  >
    <h1>
      Rock, Paper & Scissors
    </h1>
    <div
      class="home--btns-wrp"
    >
      <button
        class="game-button game-button--full-width"
        data-testid="btn-start-game"
      >
        Start Game
      </button>
      <div
        class="home--or-label"
      >
        or
      </div>
      <button
        class="game-button game-button--full-width"
        data-testid="btn-join-game"
      >
        Join a game
      </button>
      <div
        class="home--or-label"
      >
        or
      </div>
      <button
        class="game-button game-button--full-width"
      >
        See started games
      </button>
    </div>
  </div>
</div>
`;
